// assistant-stream/dist/ai-sdk/language-model.js
// src/ai-sdk/language-model.ts
import { AssistantTransformStream } from "../core/utils/stream/AssistantTransformStream.js";
function bufferToBase64(buffer) {
  return btoa(String.fromCharCode(...buffer));
}
var LanguageModelV1StreamDecoder = class extends AssistantTransformStream {
  constructor() {
    let currentToolCall;
    const endCurrentToolCall = () => {
      if (!currentToolCall) return;
      currentToolCall.controller.argsText.close();
      currentToolCall.controller.close();
      currentToolCall = void 0;
    };
    super({
      transform(chunk, controller) {
        const { type } = chunk;
        if (type === "text-delta" || type === "reasoning") {
          endCurrentToolCall();
        }
        switch (type) {
          case "text-delta": {
            controller.appendText(chunk.textDelta);
            break;
          }
          case "reasoning": {
            controller.appendReasoning(chunk.textDelta);
            break;
          }
          case "source": {
            controller.appendSource({
              type: "source",
              ...chunk.source
            });
            break;
          }
          case "file": {
            controller.appendFile({
              type: "file",
              mimeType: chunk.mimeType,
              data: typeof chunk.data === "string" ? chunk.data : bufferToBase64(chunk.data)
            });
            break;
          }
          case "tool-call-delta": {
            const { toolCallId, toolName, argsTextDelta } = chunk;
            if (currentToolCall?.toolCallId === toolCallId) {
              currentToolCall.controller.argsText.append(argsTextDelta);
            } else {
              endCurrentToolCall();
              currentToolCall = {
                toolCallId,
                controller: controller.addToolCallPart({
                  toolCallId,
                  toolName
                })
              };
              currentToolCall.controller.argsText.append(argsTextDelta);
            }
            break;
          }
          case "tool-call": {
            const { toolCallId, toolName, args } = chunk;
            if (currentToolCall?.toolCallId === toolCallId) {
              currentToolCall.controller.argsText.close();
            } else {
              const toolController = controller.addToolCallPart({
                toolCallId,
                toolName,
                argsText: args
              });
              toolController.close();
            }
            break;
          }
          case "finish": {
            controller.enqueue({
              type: "message-finish",
              finishReason: chunk.finishReason,
              usage: chunk.usage,
              path: []
            });
            controller.close();
            break;
          }
          case "error":
          case "response-metadata":
          case "reasoning-signature":
          case "redacted-reasoning":
            break;
          default: {
            const unhandledType = type;
            throw new Error(`Unhandled chunk type: ${unhandledType}`);
          }
        }
      },
      flush() {
        endCurrentToolCall();
      }
    });
  }
};
export {
  LanguageModelV1StreamDecoder
};
